lib/cache/cache.go

```diff
@@ -2491,6 +2516,43 @@ func (c *Cache) GetAccessList(ctx context.Context, name string) (*accesslist.Acc
 	return rg.reader.GetAccessList(ctx, name)
 }
 
+// GetUserLoginStates returns the all user login state resources.
+func (c *Cache) GetUserLoginStates(ctx context.Context) ([]*userloginstate.UserLoginState, error) {
+	ctx, span := c.Tracer.Start(ctx, "cache/GetUserLoginStates")
+	defer span.End()
+
+	rg, err := readCollectionCache(c, c.collections.userLoginStates)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer rg.Release()
+	return rg.reader.GetUserLoginStates(ctx)
+}
+
+// GetUserLoginState returns the specified user login state resource.
+func (c *Cache) GetUserLoginState(ctx context.Context, name string) (*userloginstate.UserLoginState, error) {
+	ctx, span := c.Tracer.Start(ctx, "cache/GetUserLoginState")
+	defer span.End()
+
+	rg, err := readCollectionCache(c, c.collections.userLoginStates)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+	defer rg.Release()
+
+	uls, err := rg.reader.GetUserLoginState(ctx, name)
+	if trace.IsNotFound(err) && rg.IsCacheRead() {
+		// release read lock early
+		rg.Release()
```

Comment:
huh, I guess we already double release the read guard in other places in this file, but it's not thread-safe. Idk the implications of this, like if another thread acquires the readlock and then our deferred release releases it...

I'd rather we just put the defer after this error checking branch